File: 2015/day-21/day21.py
#!/usr/bin/python
import re
import random
import string
from itertools import combinations

WEAPONS = [{"name": "Dagger", "cost": 8, "dmg": 4, "armor": 0},
    {"name": "Shortsword", "cost": 10, "dmg": 5, "armor": 0},
    {"name": "Warhammer", "cost": 25, "dmg": 6, "armor": 0},
    {"name": "Longsword", "cost": 40, "dmg": 7, "armor": 0},
    {"name": "Greataxe", "cost": 74, "dmg": 8, "armor": 0}]

ARMOR = [{"name": "Leather", "cost": 13, "dmg": 0, "armor": 1},
    {"name": "Chainmail", "cost": 31, "dmg": 0, "armor": 2},
    {"name": "Splintmail", "cost": 53, "dmg": 0, "armor": 3},
    {"name": "Bandedmail", "cost": 75, "dmg": 0, "armor": 4},
    {"name": "Platemail", "cost": 102, "dmg": 0, "armor": 5}]
    
RINGS = [{"name": "Damage +1", "cost": 25, "dmg": 1, "armor": 0},
    {"name": "Damage +2", "cost": 50, "dmg": 2, "armor": 0},
    {"name": "Damage +3", "cost": 100, "dmg": 3, "armor": 0},
    {"name": "Defense +1", "cost": 20, "dmg": 0, "armor": 1},
    {"name": "Defense +2", "cost": 40, "dmg": 0, "armor": 2},
    {"name": "Defense +3", "cost": 80, "dmg": 0, "armor": 3}]

class entity:
    def __init__(self, hp, damage, armor):
        self.hp = int(hp)
        self.dmg = int(damage)
        self.armor = int(armor)
        self.items = []
        self.money = 0

    def getDamage(self):
        return self.dmg

    def getCost(self):
        return self.money

    def applyBuff(self, buff):
        for item in buff:
            self.dmg += item['dmg']
            self.armor += item['armor']
            self.money += item['cost']
            self.items.append(item['name'])

    def isAlive(self):
        if self.hp <= 0:
            return False
        else:
            return True

    def printPlayerStats(self):
        return (self.dmg, self.armor, self.items)

    def hit(self, dmg):
        self.hp -= max(1, dmg-self.armor)
        return self.isAlive()

def createPlayers():
    players = []
    comb_rings = []

    for i in range(3):
        rings_i = combinations(RINGS, i)
        for j in rings_i:
            comb_rings.append(j)
    
    # create all players with all possibilites
    for wep in WEAPONS:
        for arm in ARMOR:
            for r in comb_rings:
                p = entity(100, 0, 0)               # create a new player
                p.applyBuff([wep, arm] + list(r))   # apply all buffs from the items
                players.append(p)
            
    return players

def main():
    with open('input', 'r') as fp:
        data = fp.read()
        hp, dmg, armor = re.findall(r'(\d+)', data)

    p_list = createPlayers()
    success = []
    lost = []
    for player in p_list:
        boss = entity(hp, dmg, armor)
        b_alive, p_alive = (True, True)

        for _round in range(30):
            #player goes first, then boss
            b_alive = boss.hit(player.getDamage())
            if not b_alive and p_alive:
                success.append((player.getCost(), player.printPlayerStats()))
                break
            p_alive = player.hit(boss.getDamage())
            if not p_alive:
                lost.append((player.getCost(), player.printPlayerStats()))
                break

    print(sorted(success)[0])
    print(sorted(lost, reverse=True)[0])
        

if __name__ == '__main__':
    main()

File: CSharp/aoc/year_2015/day21/day21.cs
namespace Year2015.Day21;
using Combinatorics.Collections;
public class Day21
{
    private static Dictionary<string, Item> weapons = new Dictionary<string, Item>()
    {
        {"Dagger", new Item {Name="Dagger", Price=8, Damage=4, Armour=0}},
        {"Shortsword", new Item {Name="Shortsword", Price=10, Damage=5, Armour=0}},
        {"Warhammer", new Item {Name="Warhammer", Price=25, Damage=6, Armour=0}},
        {"Longsword", new Item {Name="Longsword", Price=40, Damage=7, Armour=0}},
        {"Greataxe", new Item {Name="Greataxe", Price=74, Damage=8, Armour=0}},
    };
    private static Dictionary<string, Item> armor = new Dictionary<string, Item>()
    {
        {"Nothing", new Item {Name="Nothing", Price=0, Damage=0, Armour=0}},
        {"Leather", new Item {Name="Leather", Price=13, Armour=1, Damage=0}},
        {"Chainmail", new Item {Name="Chainmail", Price=31, Armour=2, Damage=0}},
        {"Splintmail", new Item {Name="Splintmail", Price=53, Armour=3, Damage=0}},
        {"Bandedmail", new Item {Name="Bandedmail", Price=75, Armour=4, Damage=0}},
        {"Platemail", new Item {Name="Platemail", Price=102, Armour=5, Damage=0}},
    };
    private static Dictionary<string, Item> rings = new Dictionary<string, Item>()
    {
        {"Nothing", new Item {Name="Nothing", Price=0, Damage=0, Armour=0}},
        {"Nothing 2", new Item {Name="Nothing 2", Price=0, Damage=0, Armour=0}},
        {"Damage +1", new Item {Name="Damage +1", Price=25, Damage=1, Armour=0}},
        {"Damage +2", new Item {Name="Damage +2", Price=50, Damage=2, Armour=0}},
        {"Damage +3", new Item {Name="Damage +3", Price=100, Damage=3, Armour=0}},
        {"Defense +1", new Item {Name="Defense +1", Price=20, Armour=1, Damage=0}},
        {"Defense +2", new Item {Name="Defense +2", Price=40, Armour=2, Damage=0}},
        {"Defense +3", new Item {Name="Defense +3", Price=80, Armour=3, Damage=0}},
    };

    static TextParser<Char> parser =
        (from _ in Span.Length(12)
         from hp in Character.Digit.AtLeastOnce()
         from __ in Span.Length(9)
         from dmg in Character.Digit.AtLeastOnce()
         from ___ in Span.Length(8)
         from def in Character.Digit.AtLeastOnce()
         select (new Char(int.Parse(dmg), int.Parse(def), int.Parse(hp)))
        );

    public static void Run(string Path)
    {
        int p1 = 0;
        int p2 = 0;

        var itemComb = AllItems();

        Char boss = parser.Parse(File.ReadAllText(Path));
        Char player = new Char(0, 0, 100, "Player");

        while (p1 == 0)
        {
            foreach (var (w, a, r1, r2, c) in itemComb)
            {
                boss.Reset();
                player.Reset();
                player.ApplyBuffs((w, a, r1, r2));

                string winner = Combat(player, boss);
                if (winner == "Player" && p1 == 0)
                {
                    p1 = c;
                }

                if (winner == "Boss")
                {
                    if (c > p2)
                        p2 = c;
                }
            }
        }
        Console.WriteLine($"Part 1: {p1}\nPart 2: {p2}");
    }

    private static string Combat(Char player, Char boss)
    {
        while (true)
        {
            //player goes first
            boss.HP -= (Math.Max(0, player.Attack - boss.Armour));
            if (boss.HP <= 0)
                return player.Name;

            // boss next
            player.HP -= (Math.Max(0, boss.Attack - player.Armour));
            if (player.HP <= 0)
                return boss.Name;
        }
    }

    private static List<(Item, Item, Item, Item, int)> AllItems()
    {
        string[] weps = weapons.Keys.ToArray();
        string[] armors = armor.Keys.ToArray();
        var _rings = new Combinations<string>(rings.Keys.ToList(), 2).ToList();

        var items = new List<(Item, Item, Item, Item, int)>();

        foreach (string w in weps)
        {
            foreach (string a in armors)
            {
                foreach (List<string> v in _rings)
                {
                    int cost = weapons[w].Price;
                    cost += armor[a].Price;
                    cost += rings[v[0]].Price;
                    cost += rings[v[1]].Price;
                    items.Add((weapons[w], armor[a], rings[v[0]], rings[v[1]], cost));
                }
            }
        }
        return items.OrderBy(v => v.Item5).ToList();
    }
}

public class Item
{
    public string? Name { get; set; }
    public int Price { get; set; }
    public int Damage { get; set; }
    public int Armour { get; set; }
}

public class Char
{
    public string Name;
    public int Attack;
    private int _att;
    public int Armour;
    private int _arm;
    public int HP;
    private int _hp;

    public Char(int dmg, int def, int hp, string name = "Boss")
    {
        Name = name;
        HP = hp;
        _hp = hp;
        Armour = def;
        _arm = def;
        Attack = dmg;
        _att = dmg;
    }

    public void ApplyBuffs((Item, Item, Item, Item) items)
    {
        (Item w, Item a, Item r1, Item r2) = items;
        Attack += (w.Damage + a.Damage + r1.Damage + r2.Damage);
        Armour += (w.Armour + a.Armour + r1.Armour + r2.Armour);
    }

    public void Reset()
    {
        Attack = _att;
        Armour = _arm;
        HP = _hp;
    }
}